File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        n = len(nums)
        prefix = [1] * n
        suffix = [1] * n
        output = [] 
        
        prefix[0] = nums[0]
        suffix[n-1] = nums[n-1]
        
        for i in range(1, n):
            prefix[i] = prefix[i - 1] * nums[i]
        
        for i in range(n - 2, -1, -1):
            suffix[i] = suffix[i + 1] * nums[i]
        
        for i in range(len(nums)): 
            if 0 < i < len(nums) - 1:
                output.append(prefix[i - 1] * suffix[i + 1])
            elif i == 0: 
                output.append(suffix[i + 1])
            else:
                output.append(prefix[i - 1])
        return output        

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] output = new int[n];

        // Initialize the output array with the product of elements to the left of the current index
        output[0] = 1;
        for (int i = 1; i < n; i++) {
            output[i] = output[i - 1] * nums[i - 1];
        }

        // Multiply with the product of elements to the right of the current index
        int right = 1;
        for (int i = n - 1; i >= 0; i--) {
            output[i] = output[i] * right;
            right *= nums[i];
        }

        return output;
    }
}
